libproxy/modules/config_pacrunner.cpp

```diff
@@ -0,0 +1,122 @@
+ ******************************************************************************/
+
+#include "../extension_config.hpp"
+using namespace libproxy;
+
+#include <string.h>
+#include <dbus/dbus.h>
+
+class pacrunner_config_extension : public config_extension {
+public:
+	pacrunner_config_extension() {
+		this->conn = NULL;
+	}
+
+	~pacrunner_config_extension() {
+		if (this->conn) dbus_connection_close(this->conn);
+	}
+
+	class scoped_dbus_message {
+	public:
+		scoped_dbus_message(DBusMessage *msg) {
+			this->msg = msg;
+		}
+
+		~scoped_dbus_message() {
+			if (this->msg)
+				dbus_message_unref(msg);
+		}
+
+	private:
+		DBusMessage *msg;
+	};
+
+	vector<url> get_config(const url &dest) throw (runtime_error) {
+		// Make sure we have a valid connection with a proper match
+		DBusConnection *conn = this->conn;
+		vector<url> response;
+
+		if (!conn || !dbus_connection_get_is_connected(conn))
+		{
+			// If the connection was disconnected,
+			// close it an clear the queue
+			if (conn)
+			{
+				dbus_connection_close(conn);
+				dbus_connection_read_write(conn, 0);
+				for (DBusMessage *msg=NULL ; (msg = dbus_connection_pop_message(conn)) ; dbus_message_unref(msg)) {};
+			}
+
+			// Create a new connections
+			conn = dbus_bus_get_private(DBUS_BUS_SYSTEM, NULL);
+			this->conn = conn;
+			if (!conn)
+				throw runtime_error("Unable to set up DBus connection");
+
+			// If connection was successful, set it up
+			dbus_connection_set_exit_on_disconnect(conn, false);
+		}
+
+		DBusMessage *msg, *reply;
+
+		msg = dbus_message_new_method_call("org.pacrunner",
+						   "/org/pacrunner/client",
+						   "org.pacrunner.Client",
+						   "FindProxyForURL");
+		if (!msg)
+			throw runtime_error("Unable to create PacRunner DBus call");
+
+		string dest_str = dest.to_string();
+		string dest_host = dest.get_host();
+		const char *dest_cstr = dest_str.c_str();
+		const char *dest_host_cstr = dest_host.c_str();
+
+		dbus_message_append_args(msg, DBUS_TYPE_STRING, &dest_cstr,
+					 DBUS_TYPE_STRING, &dest_host_cstr,
+					 DBUS_TYPE_INVALID);
+
+		reply = dbus_connection_send_with_reply_and_block(conn, msg, -1, NULL);
+
+		dbus_message_unref(msg);
+
+		if (!reply)
+			throw runtime_error("Failed to get DBus response from PacRunner");
+
+		scoped_dbus_message smsg(reply);
+		char *str = NULL;
+		dbus_message_get_args(reply, NULL, DBUS_TYPE_STRING, &str, DBUS_TYPE_INVALID);
+
+		if (!str || !strlen(str) || !::strcmp(str, "DIRECT"))
+			response.push_back(url("direct://"));
+		else if (!strncmp(str, "PROXY ", 6))
+			response.push_back(url("http://" + string(str + 6)));
+		else if (!strncmp(str, "SOCKS ", 6))
+			response.push_back(url("socks://" + string(str + 6)));
+		else {
+			throw runtime_error("Unrecognised proxy response from PacRunner: " + string(str));
+		}
+		return response;
+	}
+
+private:
+	DBusConnection *conn;
+};
+
+MM_MODULE_INIT_EZ(pacrunner_config_extension, true, NULL, NULL);
```

Comment:
That sounds a bit generic to be an always valid config module... makes the module apparently pick up the config even if pacrunner is not being used.

The 'make thus module valid' test should at least query if pacrunner listens on the dbus... if not, it's very likely not in use.

@npmccallum thoughts?